docs/source/core/exp_manager.rst

```diff
@@ -322,6 +322,7 @@ and stability. To use EMA, simply set the following via YAML or :class:`~nemo.ut
 
 
 .. _nemo_multirun-label:
+
```

Comment:
I am providing line-by-line copyedits to this file here because most of the file is read only.

Line 7. revise.
The NeMo Framework Experiment Manager leverages PyTorch Lightning for model checkpointing, TensorBoard Logging, Weights and Biases, DLLogger and MLFlow logging.

Line 10. fix capitalization, punctuation, and revise (remove simply).
To use the Experiment Manager, call :class:`~nemo.utils.exp_manager.exp_manager` and pass in the PyTorch Lightning ``Trainer``.

Line 16. revise so it is a complete sentence. suggested revision.
The Experiment Manager is configurable using YAML with Hydra.

Line 36. fix punctuation.
We can configure the ``ModelCheckpoint`` via YAML or CLI:

Line 54-56. revise for clarity.
To auto-resume training, configure the exp_manager. This feature is important for long training runs that might be interrupted or shut down before the procedure has completed. To auto-resume training, set the following parameters via YAML or CLI:

Line 76. revise (remove simply).
Alongside Tensorboard, NeMo also supports Weights and Biases, MLFlow, DLLogger, ClearML and NeptuneLogger. To use these loggers, set the following
via YAML or :class:`~nemo.utils.exp_manager.ExpManagerConfig`.

Line 182 revise (remove simply) and fix punctuation.
To use EMA, set the following parameters via YAML or :class:`~nemo.utils.exp_manager.ExpManagerConfig`.

Line 204. fix punctuation and revise (remove simply).
 The ``PreemptionCallback`` is enabled by default. To disable it, add ``create_preemption_callback: False`` under exp_manager in the config YAML file. 

Line 214. suggested revision.
    Distributed training can be affected by stragglers, which are workers that slow down the overall training process. 

Lines 223-224. Change - to the bullet character * for proper formatting.

        * Relative GPU performance score: The best-performing GPU in the workload is used as a reference.
        * Individual GPU performance score: The best historical performance of the GPU is used as a reference.

Lines 226-228. fix lead-in sentence, change to bullets, and fix punctuation.

For example:

       * If the relative performance score is 0.5, it means that a GPU is twice slower than the fastest GPU.
       * If the individual performance score is 0.5, it means that a GPU is twice slower than its best observed. 

Line 249. suggested revision.
Straggler detection may require inter-rank synchronization and should be performed at regular intervals, such as every few minutes.

Line 260. Do we need to define DNN?
When training Deep Neural Network (DNN models), faults may occur, hindering the progress of the entire training process. 
    This is particularly common in distributed, multi-node training scenarios, with many nodes and GPUs involved.

Lines 295-299. fix punctuation and grammar.
 If ``calculate_timeouts: True``, timeouts will be automatically estimated based on observed intervals. 
    Estimated timeouts take precedence over timeouts defined in the config file. **Timeouts are estimated 
    at the end of a training run when checkpoint loading and saving were observed**. Hence, in a multi-part 
    training started from scratch, estimated timeouts won't be available during the initial two runs. 
    Estimated timeouts are stored in a separate JSON file.

Line 319. fix auto-resuming.
        * ``max_subsequent_job_failures`` (int, default=0) Used by FT launcher. How many subsequent job failures are allowed until stopping auto-resuming.

Line 320. fix auto-resume. 
``0`` means do not auto-resume. 

Lines 329-322. suggested revision.

When training neural networks, it is common to perform a hyperparameter search to improve the model’s performance on validation data. However, manually preparing a grid of experiments and managing all checkpoints and their metrics can be tedious. To simplify these tasks, NeMo integrates with Hydra Multi-Run support, providing a unified way to run a set of experiments directly from the configuration.

Lines 334-338. suggested revision.
There are certain limitations to this framework, which are listed below:

* All experiments are assumed to be run on a single GPU. Multi-GPU for a single run (model parallel models) is not supported at this time.
* NeMo Multi-Run currently supports only grid search over a set of hyperparameters. Support for advanced hyperparameter search strategies will be added in the future.
* NeMo Multi-Run requires one or more GPUs to function and will not work without GPU devices.


Lines 340-341. Change title to imperative verb. suggested revision.

Configure NeMo Multi-Run
~~~~~~~~~~~~~~~~~~~~~~~~

Line 343. suggested revision.
To enable NeMo Multi-Run, we first update our YAML configs with the necessary information to inform Hydra that we expect to run multiple experiments from this single configuration.

Lines 373-376. suggested revision.
Next, we will set up the configuration for the Experiment Manager. When performing a hyperparameter search, each run may take some time to complete. To avoid redoing all experiments in case a run ends prematurely (for example, due to an out-of-memory error or a timeout on the machine), we configure the Experiment Manager such that every experiment has a unique “key”. This key corresponds to a single resumable experiment.

Lines 378-379. revise.
Let's see how to set up such a unique "key" using the experiment name. To do this, attach all the hyperparameter arguments to the experiment name as shown below.

Line 409. revise title.

Run a NeMo Multi-Run Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Line 410. fix punctuation.
Once the config has been updated, we can now run it just like any normal Hydra script, with one special flag ``(-m)!``.

Line 422. need to add some introductory text for this section. suggested addition.
This section provides recommendations for using the Experiment Manager.

Line 423. Change this tip to a heading so it stands out.

Preserving Disk Space for a Large Number of Experiments
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Lines 425-427 suggested revision.

Some models may have a large number of parameters, making it very expensive to save numerous checkpoints on physical storage drives. For example, if you use the Adam optimizer, each PyTorch Lightning “.ckpt” file will be three times the size of just the model parameters per checkpoint file. This can become exorbitant if you have multiple runs.

Lines 429-431. suggested revision.
In the above configuration, we explicitly set save_top_k: 1 and always_save_nemo: True. This limits the number of checkpoint files to just one and also saves a NeMo file, which contains only the model parameters without the optimizer state. This NeMo file can be restored immediately for further work.

Lines 433-435 suggested revision.

We can further reduce storage space by using NeMo’s utility functions to automatically delete either checkpoint or NeMo files after a training run has finished. This is sufficient if you are collecting results in an experiment tracking tool and can simply rerun the best configuration after the search is completed.


Line 456. Change this tip to a heading so it stands out.

Debugging Multi-Run Scripts
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Lines 458-461. suggested revision.

When running Hydra scripts, you may encounter configuration issues that crash the program. In NeMo Multi-Run, a crash in any single run will not crash the entire program. Instead, we will note the error and proceed to the next job. Once all jobs are completed, we will raise the errors in the order they occurred, crashing the program with the first error’s stack trace.

Lines 463-464. suggested revision.

To debug NeMo Multi-Run, we recommend commenting out the entire hyperparameter configuration set inside sweep.params. Instead, run a single experiment with the configuration, which will immediately raise the error.

Line 467. Change this tip to a heading so it stands out.

Experiment Name Cannot be Parsed by Hydra
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Lines 469-471. suggested revision.

Sometimes our hyperparameters include PyTorch Lightning trainer arguments, such as the number of steps, number of epochs, and whether to use gradient accumulation. When we attempt to add these as keys to the experiment manager’s name, Hydra may complain that trainer.xyz cannot be resolved.

Line 473. fix capitalization and punctuation.
A simple solution is to finalize the Hhydra config before you call ``exp_manager()`` as follows: